File: dotNET/PFProSDK/src/PayPal/Payments/Communication/SendState.cs
#region "Copyright"

//PayPal Payflow Pro .NET SDK
//Copyright (C) 2014  PayPal, Inc.
//
//This file is part of the Payflow Pro .NET SDK
//
//The Payflow .NET SDK is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//any later version.
//
//The Payflow .NET SDK is is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with the Payflow .NET SDK.  If not, see <http://www.gnu.org/licenses/>.


#endregion

#region "Imports"

using System;
using PayPal.Payments.Common.Logging;
using PayPal.Payments.Common;
using PayPal.Payments.Common.Utility;

#endregion

namespace PayPal.Payments.Communication
{
	/// <summary>
	/// Represents SendState.
	/// </summary>
	abstract internal class SendState : PaymentState
	{
		#region "Constructors"

		/// <summary>
		/// Copy constructor for SendState.
		/// </summary>
		/// <param name="CurrentPaymentState">Current Payment State object.</param>
		public SendState(PaymentState CurrentPaymentState) : base(CurrentPaymentState)
		{
		}

		#endregion		

		#region "Functions"

		/// <summary>
		/// Abstract Declaration of 
		/// GetSendRequest
		/// </summary>
		/// <returns>request to be sent</returns>
		public abstract String GetSendRequest();

		/// <summary>
		/// Execute function
		/// </summary>
		public override void Execute()
		{
			bool IsSendSuccess = false;

			Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Entered.", PayflowConstants.SEVERITY_DEBUG);
			if (!InProgress)
				return;
			try
			{
				//Begin Payflow Timeout Check Point 3
				long TimeRemainingMsec;
				if (PayflowUtility.IsTimedOut(mConnection.TimeOut, mConnection.StartTime, out TimeRemainingMsec))
				{
					String AddlMessage = "Input timeout value in millisec = " + Connection.TimeOut.ToString();
					ErrorObject Err = PayflowUtility.PopulateCommError(PayflowConstants.E_TIMEOUT_WAIT_RESP, null, PayflowConstants.SEVERITY_FATAL, IsXmlPayRequest, AddlMessage);
					if (!CommContext.IsCommunicationErrorContained(Err))
					{
						CommContext.AddError(Err);
					}
				}
				else
				{
					mConnection.TimeOut = TimeRemainingMsec;
				}
				//End Payflow Timeout Check Point 3
				IsSendSuccess = mConnection.SendToServer(GetSendRequest());
			}
			catch (Exception Ex)
			{
				Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Error occurred While Initializing Connection.", PayflowConstants.SEVERITY_ERROR);
				Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Exception " + Ex.ToString(), PayflowConstants.SEVERITY_ERROR);
				IsSendSuccess = false;
			}
            //catch
            //{
            //    IsSendSuccess = false;
            //}
			finally
			{
				if (IsSendSuccess)
				{
					Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Send Data =  Success ", PayflowConstants.SEVERITY_INFO);
					SetStateSuccess();
				}
				else
				{
					Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Send Data =  Failure ", PayflowConstants.SEVERITY_INFO);
					SetStateFail();
				}
			}
			Logger.Instance.Log("PayPal.Payments.Communication.SendState.Execute(): Exiting.", PayflowConstants.SEVERITY_DEBUG);
		}

		#endregion
	}
}

File: java/src/sdk/base/paypal/payflow/SendState.java
package paypal.payflow;

/*
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Represents SendState.
 */
abstract class SendState extends PaymentState {

    /**
     * Copy constructor for SendState.
     *
     * @param currentPaymentState PaymentState
     */

    public SendState(PaymentState currentPaymentState) {
        super(currentPaymentState);
        Logger.getInstance().log("paypal.payflow.SendState(PaymentState) : Entered", PayflowConstants.SEVERITY_DEBUG);
        Logger.getInstance().log("paypal.payflow.SendState(PaymentState) : Exiting", PayflowConstants.SEVERITY_DEBUG);
    }


    /**
     * Abstract Declaration of
     * GetSendRequest
     *
     * @return Return
     */

    public abstract String getSendRequest();

    /**
     * Execute function
     */
    public void execute() {
        boolean isSendSuccess = false;

        Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Entered", PayflowConstants.SEVERITY_DEBUG);
        if (getInProgress()) {
            try {
                //Begin Payflow Timeout Check Point 3
                if (PayflowUtility.isTimedOut(mConnection.getTimeout(), mConnection.getStartTime())) {
                    String addlMessage = "Input timeout value in millisec = " + getConnection().getTimeout();
                    ErrorObject err = PayflowUtility.populateCommError(PayflowConstants.E_TIMEOUT_WAIT_RESP, null, PayflowConstants.SEVERITY_FATAL, getIsXmlPayRequest(), addlMessage);
                    if (!getCommContext().isCommunicationErrorContained(err)) {
                        getCommContext().addError(err);
                    }
                }
                //End Payflow Timeout Check Point 3
                isSendSuccess = mConnection.sendToServer(getSendRequest());
            } catch (Exception ex) {
                Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Following Error occurred While Initializing Connection.", PayflowConstants.SEVERITY_ERROR);
                Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Exception " + ex.toString(), PayflowConstants.SEVERITY_ERROR);
                isSendSuccess = false;
            } finally {
                if (isSendSuccess) {
                    Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Send Data =  Success ", PayflowConstants.SEVERITY_INFO);
                    setStateSuccess();
                } else {
                    Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Send Data =  Failure ", PayflowConstants.SEVERITY_INFO);
                    setStateFail();
                }
            }
        }

        Logger.getInstance().log("paypal.payflowCommunication.SendState.Execute(): Exiting", PayflowConstants.SEVERITY_DEBUG);
    }


}
